crates/ra_hir_ty/src/_match.rs

```diff
@@ -0,0 +1,1174 @@
+#[derive(Debug)]
+/// Similar to TypeCtor, but includes additional information about the specific
+/// value being instantiated. For example, TypeCtor::Bool doesn't contain the
+/// boolean value.
+enum Constructor {
+    Bool(bool),
+    Tuple { arity: usize },
+    Enum(EnumVariantId),
+}
+
+/// Returns the constructor for the given pattern. Should only return None
+/// in the case of a Wild pattern.
+fn pat_constructor(cx: &MatchCheckCtx, pat: PatIdOrWild) -> MatchCheckResult<Option<Constructor>> {
+    let res = match pat.as_pat(cx) {
+        Pat::Wild => None,
+        Pat::Tuple(pats) => Some(Constructor::Tuple { arity: pats.len() }),
+        Pat::Lit(lit_expr) => match cx.body.exprs[lit_expr] {
+            Expr::Literal(Literal::Bool(val)) => Some(Constructor::Bool(val)),
+            _ => return Err(MatchCheckNotImplemented),
+        },
+        Pat::TupleStruct { .. } | Pat::Path(_) => {
+            let pat_id = pat.as_id().expect("we already know this pattern is not a wild");
+            let variant_id =
+                cx.infer.variant_resolution_for_pat(pat_id).ok_or(MatchCheckNotImplemented)?;
+            match variant_id {
+                VariantId::EnumVariantId(enum_variant_id) => {
+                    Some(Constructor::Enum(enum_variant_id))
+                }
+                _ => return Err(MatchCheckNotImplemented),
+            }
+        }
+        _ => return Err(MatchCheckNotImplemented),
+    };
+
+    Ok(res)
+}
+
+fn all_constructors_covered(
+    cx: &MatchCheckCtx,
+    constructor: &Constructor,
+    used_constructors: &[Constructor],
+) -> bool {
+    match constructor {
+        Constructor::Tuple { arity } => {
+            used_constructors.iter().any(|constructor| match constructor {
+                Constructor::Tuple { arity: used_arity } => arity == used_arity,
+                _ => false,
+            })
+        }
+        Constructor::Bool(_) => {
+            if used_constructors.is_empty() {
+                return false;
+            }
+
+            let covers_true =
+                used_constructors.iter().any(|c| matches!(c, Constructor::Bool(true)));
+            let covers_false =
+                used_constructors.iter().any(|c| matches!(c, Constructor::Bool(false)));
+
+            covers_true && covers_false
+        }
+        Constructor::Enum(e) => cx.db.enum_data(e.parent).variants.iter().all(|(id, _)| {
+            for constructor in used_constructors {
+                if let Constructor::Enum(e) = constructor {
+                    if id == e.local_id {
+                        return true;
+                    }
+                }
+            }
+
+            false
+        }),
+    }
+}
+
+fn enum_variant_matches(cx: &MatchCheckCtx, pat_id: PatId, enum_variant_id: EnumVariantId) -> bool {
+    if let Some(VariantId::EnumVariantId(pat_variant_id)) =
+        cx.infer.variant_resolution_for_pat(pat_id)
+    {
+        if pat_variant_id.local_id == enum_variant_id.local_id {
```

Comment:
there's no need to only compare the local ids, so how about just `cx.infer.variant_resolution_for_pat(pat_id) == Some(enum_variant_id)` for the whole method?

Reply:
Cool, thanks for pointing this out :+1: